File: 20221016_Sunday/20221016.rb
# Leetcode: 1335. Minimum Difficulty of a Job Schedule.
# https://leetcode.com/problems/minimum-difficulty-of-a-job-schedule/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 1558 ms, faster than 83.33% of Ruby online submissions for Minimum Difficulty of a Job Schedule.
# Memory Usage: 211 MB, less than 100.00% of Ruby online submissions for Minimum Difficulty of a Job Schedule.
# 2022.10.16 Daily Challenge.
# @param {Integer[]} job_difficulty
# @param {Integer} d
# @return {Integer}
def min_difficulty(job_difficulty, d)
  # Based on:
  # https://leetcode.com/problems/minimum-difficulty-of-a-job-schedule/discuss/490316/JavaC%2B%2BPython3-DP-O(nd)-Solution
  n, inf = job_difficulty.size, Float::INFINITY
  dp = Array.new(d+1).map{ Array.new(n, Float::INFINITY) + [0]}
  (1..d).each do |de|
    (0..n-de).each do |i|
      maxd = 0
      (i..n-de).each do |j|
        maxd = [maxd, job_difficulty[j]].max
        dp[de][i] = [dp[de][i],maxd + dp[de-1][j+1]].min
      end
    end
  end
  dp[d][0] < inf ? dp[d][0] : -1
end

File: 20221016_Sunday/20221016.go
// Leetcode: 1335. Minimum Difficulty of a Job Schedule.
// https://leetcode.com/problems/minimum-difficulty-of-a-job-schedule/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 12 ms, faster than 76.74% of Go online submissions for Minimum Difficulty of a Job Schedule.
// Memory Usage: 2.5 MB, less than 83.72% of Go online submissions for Minimum Difficulty of a Job Schedule.
// 2022.10.16. Daily Challenge.
func minDifficulty(jobDifficulty []int, d int) int {
  // Based on:
  // https://leetcode.com/problems/minimum-difficulty-of-a-job-schedule/discuss/490316/JavaC%2B%2BPython3-DP-O(nd)-Solution
  n, inf := len(jobDifficulty), 4294967295
  dp := make([][]int, d+1)
  fill_slice(dp, d+1, n+1, inf)
  for de := 1; de <= d ; de +=1 {
    for i := 0; i <= n-de; i +=1{
      maxd := 0
      for j := i; j <= n-de; j += 1{
        maxd = max(maxd, jobDifficulty[j])
        dp[de][i] = min(dp[de][i], maxd + dp[de-1][j+1])
      }
    }
  }
  if dp[d][0] < inf {
    return dp[d][0]
  }
  return -1
}

func fill_slice(slice [][]int, rows int, cols int, value int){
  for i:=0; i < rows; i += 1{
    slice[i] = make([]int, cols)
    for j :=0 ; j < cols; j += 1{
      if j == cols - 1 {
        slice[i][j] = 0
      }else{
        slice[i][j] = value 
      }
    }
  }
}
func max(a int,b int) int{
  if a > b {
    return a
  }
  return b
}
func min(a int, b int) int{
  if a < b {
    return a
  }
  return b
}
